tests/integration/ha_tests/application-charm/src/charm.py

```diff
@@ -128,33 +143,26 @@ def _start_continuous_writes(self, starting_number: int) -> None:
         )
 
         # Store the continuous writes process ID to stop the process later.
-        self._stored.continuous_writes_pid = popen.pid
+        self.app_peer_data[PROC_PID_KEY] = str(popen.pid)
 
     def _stop_continuous_writes(self) -> Optional[int]:
         """Stops continuous writes to PostgreSQL and returns the last written value."""
-        if self._stored.continuous_writes_pid is None:
+        if not self.app_peer_data.get(PROC_PID_KEY):
             return None
 
         # Stop the process.
-        proc = subprocess.Popen(["pkill", "--signal", "SIGKILL", "-f", "src/continuous_writes.py"])
-
-        # Wait for process to be killed.
-        proc.communicate()
+        os.kill(int(self.app_peer_data[PROC_PID_KEY]), signal.SIGTERM)
 
-        self._stored.continuous_writes_pid = None
+        del self.app_peer_data[PROC_PID_KEY]
```

Comment:
`os.kill()` failed when using the stored state PID, so copied over the logic from the mongodb-k8s and mysql-k8s charms. This might have also caused issues with the suite, if the PID being killed was wrong.

Reply:
Great idea!